File: dotnet-desktop-guide/samples/snippets/visualbasic/VS_Snippets_Winforms/System.Windows.Forms.WebBrowser.ObjectForScripting/vb/form1.vb
'<snippet0>
Imports System.Windows.Forms

Public Class Form1
    Inherits Form

    Private webBrowser1 As New WebBrowser()
    Private WithEvents button1 As New Button()

    <STAThread()> _
    Public Shared Sub Main()
        Application.EnableVisualStyles()
        Application.Run(New Form1())
    End Sub

    Public Sub New()
        button1.Text = "call script code from client code"
        button1.Dock = DockStyle.Top
        webBrowser1.Dock = DockStyle.Fill
        Controls.Add(webBrowser1)
        Controls.Add(button1)
    End Sub

    Protected Overrides Sub OnLoad(e As EventArgs)
        MyBase.OnLoad(e)

        '<snippet1>
        webBrowser1.AllowWebBrowserDrop = False
        '</snippet1>
        '<snippet2>
        webBrowser1.IsWebBrowserContextMenuEnabled = False
        '</snippet2>
        '<snippet3>
        webBrowser1.WebBrowserShortcutsEnabled = False
        '</snippet3>
        '<snippet4>
        webBrowser1.ObjectForScripting = New MyScriptObject(Me)
        '</snippet4>
        '<snippet9>
        ' Uncomment the following line when you are finished debugging.
        'webBrowser1.ScriptErrorsSuppressed = True
        '</snippet9>

        webBrowser1.DocumentText = _
            "<html><head><script>" & _
            "function test(message) { alert(message); }" & _
            "</script></head><body><button " & _
            "onclick=""window.external.Test('called from script code')"" > " & _
            "call client code from script code</button>" & _
            "</body></html>"
    End Sub

    Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) _
        Handles button1.Click

        '<snippet8>
        webBrowser1.Document.InvokeScript("test", _
            New String() {"called from client code"})
        '</snippet8>

    End Sub

End Class

'<snippet5>
Public Class MyScriptObject
    Private _form As Form1

    Public Sub New(ByVal form As Form1)
        _form = form
    End Sub

    Public Sub Test(ByVal message As String)
        MessageBox.Show(message, "client code")
    End Sub

End Class
'</snippet5>
'</snippet0>

File: dotnet-desktop-guide/samples/snippets/csharp/VS_Snippets_Winforms/System.Windows.Forms.WebBrowser.ObjectForScripting/CS/form1.cs
//<snippet0>
using System;
using System.Windows.Forms;

public class Form1 : Form
{
    private WebBrowser webBrowser1 = new WebBrowser();
    private Button button1 = new Button();

    [STAThread]
    public static void Main()
    {
        Application.EnableVisualStyles();
        Application.Run(new Form1());
    }

    public Form1()
    {
        button1.Text = "call script code from client code";
        button1.Dock = DockStyle.Top;
        button1.Click += new EventHandler(button1_Click);
        webBrowser1.Dock = DockStyle.Fill;
        Controls.Add(webBrowser1);
        Controls.Add(button1);
    }

    protected override void OnLoad(EventArgs e)
    {
        base.OnLoad(e);

        //<snippet1>
        webBrowser1.AllowWebBrowserDrop = false;
        //</snippet1>
        //<snippet2>
        webBrowser1.IsWebBrowserContextMenuEnabled = false;
        //</snippet2>
        //<snippet3>
        webBrowser1.WebBrowserShortcutsEnabled = false;
        //</snippet3>
        //<snippet4>
        webBrowser1.ObjectForScripting = new MyScriptObject(this);
        //</snippet4>
        //<snippet9>
        // Uncomment the following line when you are finished debugging.
        //webBrowser1.ScriptErrorsSuppressed = true;
        //</snippet9>

        webBrowser1.DocumentText =
            "<html><head><script>" +
            "function test(message) { alert(message); }" +
            "</script></head><body><button " +
            "onclick=\"window.external.Test('called from script code')\">" +
            "call client code from script code</button>" +
            "</body></html>";
    }

    private void button1_Click(object sender, EventArgs e)
    {
        //<snippet8>
        webBrowser1.Document.InvokeScript("test",
            new String[] { "called from client code" });
        //</snippet8>
    }
}

//<snippet5>
public class MyScriptObject
{
    private Form1 _form;

    public MyScriptObject(Form1 form)
    {
        _form = form;
    }

    public void Test(string message)
    {
        MessageBox.Show(message, "client code");
    }
}
//</snippet5>
//</snippet0>